### ISSUE TITLE
Tune pre Neural Network

### ISSUE THREAD

**Issue by 14corman** [14corman/JMIProject#7]
Try to tune the pre neural network to get the best test accuracy/F1 score. 

This could mean:
1. making new hidden layers
2. adjusting the number of neurons in each hidden layer
3. changing learning rate
4. changing drop out percentage
...

**Comment by 14corman** [comment]
![image](https://user-images.githubusercontent.com/15676385/53689441-b1541600-3d1b-11e9-9073-5ab1b738df1d.png)



![image](https://user-images.githubusercontent.com/15676385/53689436-8d90d000-3d1b-11e9-8b2c-8c012b1b9e78.png)


![image](https://user-images.githubusercontent.com/15676385/53689429-66d29980-3d1b-11e9-80b6-40788e1f8a33.png)


After testing for a few hours yesterday, these are some of the results I was seeing on tensorboard. The longer lines ran for 170 epochs. The pink line was not only the best, but it required very little training to achieve that. I would suspect that with the data we are using, the best we can achieve, right now speaking, is the pink line. I might try and see if my desktop can run to 500 epochs tomorrow with a few more trials to see if I can get better results, but for right now this is what we have. The labels for the graphs show the date and time along with the type of model I used with the learning rate. 

I will be uploading these models as well the updated code to create them in a few minutes.

**Comment by 14corman** [comment]
New findings. I am still doing tests, but so far I have made 5 different sized models. I test each model size against:
drop out %: [0.0, 0.2, 0.4, 0.6, 0.8, 1]
learning rate: [0.0001, 0.001, 0.01, 0.1]
So, each model is trained and evaluated 24 different times automatically through 2 for loops, and I just need to input the model size. The graphs I will be putting here have the top model's results from each size.

![image](https://user-images.githubusercontent.com/15676385/53746298-81655980-3e66-11e9-8685-3ebfe581b7a9.png)


![image](https://user-images.githubusercontent.com/15676385/53746336-95a95680-3e66-11e9-9289-969fcb0389fa.png)


![image](https://user-images.githubusercontent.com/15676385/53746372-a8bc2680-3e66-11e9-9364-6a503cf2f33f.png)


For each best model, I am looking at its data in a csv file that tensorflow has made, and am graphing the models loss against its validation loss. Since both loss values, so far, have been going down together, that means that we are not over-fitting yet and that we can make the models larger which is what I am doing. The maximum the models can be trained for is 500 epochs, but I have a "plateau" callback going into the model that stops training when the validation f1 score has  plateaued for 50 epochs.

**Comment by 14corman** [comment]
The final results are in. I did a very large network last night (1000, 2000, 2000, 1000). From there I got the best model, and these graphs show it alongside the other models' best performers. The top dark blue marker in the legend is the biggest model.

![image](https://user-images.githubusercontent.com/15676385/53807239-ab755500-3f14-11e9-9a44-73347a9535b7.png)


![image](https://user-images.githubusercontent.com/15676385/53807277-c778f680-3f14-11e9-9317-b6e82a439fca.png)


![image](https://user-images.githubusercontent.com/15676385/53807319-e37c9800-3f14-11e9-93a2-dc779d68dba2.png)

This new model is best by a decent margin with a validation accuracy of nearly 80%, and an f1 score of about .8. I wonder if making a bigger model may help, and I wonder why all the best models either drop nothing or drop everything (0.0 or 1.0). I am not sure if Keras has a minus epsilon so the top one is 0.99999999 or something because if it was truly 1 then no nodes would be used. I almost put that in as a test to see if the model could learn at all. I did not expect it to actually be best hyper parameter sets for the models.

**Comment by 14corman** [comment]
The large network that I talked about in my previous comment was over 300 MB in size, so I could not push that to Github. That will stay on my desktop at home for now.

**Comment by 14corman** [comment]
So I just read up on Keras Dropout layer. People say if you put in a dropout rate of either 0.0 or 1.0 then the layer does nothing, but if you put in a dropout of 0.99 then it drops everything. Good to know, all the 1.0 rates that I have are effectively dropping nothing, and we are doing 2 different training times with no drop out.

Here is the actual line of code in Keras:
``` python 
if 0. < self.rate < 1.
```


More info can be found here:
[https://stackoverflow.com/questions/48355700/dropout-behavior-in-keras-with-rate-1-dropping-all-input-units-not-as-expected](https://stackoverflow.com/questions/48355700/dropout-behavior-in-keras-with-rate-1-dropping-all-input-units-not-as-expected)

**Comment by 14corman** [comment]
I just put the best result from all layer batch norm in with the front layer batch norm and no batch norm for the model with layers 1000 2000 2000 10000. The orange line has no normalization, the has all layers normalized, and the blue has the first layer normalized.

![image](https://user-images.githubusercontent.com/15676385/54065312-dbf31280-41e4-11e9-8371-d7651073db44.png)

![image](https://user-images.githubusercontent.com/15676385/54065364-74899280-41e5-11e9-9d0c-d730247059ff.png)


From these, it is clear to see that we do not need regularization yet. We are underfitting the data which is why regularization techniques are not working yet (drop out and normalization). The only things that could help are a large network, or more data. I am about at the limit for the network size on my gtx 1080, and I do not think we can make fake data points since the 3 input variables are very rigid and have a small range. With this, I am allowing all 0's, 1's, and 3's to be put in, and I have the base number of 2's. I want to see  if putting all the data in can get us closer to what we need first. If this seems to work, then I will see if I can randomly add between 0 and (maybe 3-5) more 2 data points to artificially increase them. However, since I cannot add noise to them I feel that they may overfit 2's. We will see what happens. This all includes the batch normalization on all layers. I will let everyone know what happens when these 2 finish.


Oh the number of training data points is going up from about 12,000 to 72,000. Hopefully that is more than enough.

**Comment by 14corman** [comment]
Great news, my computer finally stopped training, or at least I forced it to stop. It was training model 22 out of 24, but those models are the dropout = 1.0 which we now know is = dropout = 0.0. I looked at the results and since 0.0 did ok but not the best, I decided to stop there. Training took forever with 34 hours of it. Our best 2 models out of this either had dropout = 0.2 or 0.4. 0.2 was just barely better than 0.4, so I decided to take that. They both had a learning rate of 0.001. Here are pictures of its performance compared with the tests I showed in the last message:

![image](https://user-images.githubusercontent.com/15676385/54084597-b3603b00-4300-11e9-9a10-c22bc2a96d5e.png)


![image](https://user-images.githubusercontent.com/15676385/54084602-ce32af80-4300-11e9-9b7d-0528db6ab20d.png)


![image](https://user-images.githubusercontent.com/15676385/54084611-eb677e00-4300-11e9-83d8-e59ce1ddaa2c.png)



The light blue line, marked all data, is the  newest training that I mentioned above. There are a few things to point out from those graphs:
1. It did not need to train for nearly as many epochs, but each epoch took 36 seconds rather than 5 second.
2. It started to overfit the data as shown by loss starting to go up and f1 and accuracy going down.
3. It did dramatically better than the trials with little data even though 2 was a very underrepresented class.

The thing to remember is that the epoch with the best f1 score was saved, so we do not have the ending of the line, but the peak. We almost got 90% validation accuracy! 

I want to try 1 more test before we say this is the best of the best of what we can get. I am going to randomly add more 2's to the dataset that is returned. More precisely, when a 2 is hit, it will pick a number between 0 and 3 (or 4) and add that many more 2's to the returning dataset. This would add more data, and not have 2's be as underrepresented. However, my fear is that since I am not adding any noise to the added data points that the model will overfit 2's to those data points. We will see what this holds. For now, our best model is:
1. layers: 1000, 2000, 2000, 10000
2. dropout: 0.2
3. learning rate: 0.001
4. Batch normalization as each layer
5. Include all data.

**Comment by 14corman** [comment]
Actually, I think we would need more data in general. I just ran the predict.py file on the model with 20 and 200 predictions. It got 70% and 69% accuracy respectively. That is pretty good for a classifier. I think adding 2's will not work as well as we would hope. The only thing that could help this model is more data in general, but more specifically more 2 label data. I think I will leave this here for now until analysis is done.
